File: runtimes/http-ts/test/retry.test.ts
import { afterEach, describe, expect, it, vi } from "vitest";

import type {
  CanonicalResponse,
  CanonicalTransport,
  Outcome,
  WireDescriptor,
} from "../src/descriptor";
import { assertExclusiveTransport, execute } from "../src/execute";
import {
  backoffDelayMs,
  isRetryable,
  resolveMaxRetries,
  resolveTimeoutMs,
} from "../src/retry";

afterEach(() => {
  vi.useRealTimers();
  vi.restoreAllMocks();
});

function descriptor(over: Partial<WireDescriptor> = {}): WireDescriptor {
  return {
    http_method: "POST",
    uri: "/things",
    bindings: [],
    response_bindings: [],
    success: [[200, null]],
    ...over,
  };
}

// Deterministic seams: jitter pinned to 0.5 and a recording, instant sleep.
function seams() {
  const delays: number[] = [];
  return {
    delays,
    seams: {
      sleep: (ms: number) => {
        delays.push(ms);
        return Promise.resolve();
      },
      random: () => 0.5,
    },
  };
}

// A canonical transport answering from a script, one step per attempt.
type Step =
  | { kind: "response"; status: number; body: string }
  | { kind: "transport_failure" }
  | { kind: "hang" };

function scripted(steps: Step[]) {
  let attempts = 0;
  const transport: CanonicalTransport = (_req, signal) => {
    if (attempts >= steps.length) {
      return Promise.reject(new Error(`transport called ${attempts + 1} times for a ${steps.length}-step script`));
    }
    const step = steps[attempts];
    attempts += 1;
    if (step.kind === "response") {
      return Promise.resolve({ status: step.status, headers: {}, body: step.body });
    }
    if (step.kind === "transport_failure") {
      return Promise.reject(new Error("scripted transport failure"));
    }
    // Honor the abort signal, as the canonical transport contract asks; bail
    // out fast if a mutant keeps the timeout from firing.
    return new Promise<CanonicalResponse>((_, reject) => {
      const bail = setTimeout(() => reject(new Error("hang step: the per-attempt timeout never fired")), 2000);
      signal?.addEventListener("abort", () => {
        clearTimeout(bail);
        reject(signal.reason instanceof Error ? signal.reason : new Error("aborted"));
      });
    });
  };
  return { transport, attempts: () => attempts };
}

describe("value resolution", () => {
  it("resolves the retry maximum from a literal, flooring and disabling below one", () => {
    expect(resolveMaxRetries(undefined, undefined)).toBe(0);
    expect(resolveMaxRetries(null, undefined)).toBe(0);
    expect(resolveMaxRetries({ max: { lit: 2 } }, undefined)).toBe(2);
    expect(resolveMaxRetries({ max: { lit: 2.9 } }, undefined)).toBe(2);
    expect(resolveMaxRetries({ max: { lit: 0.5 } }, undefined)).toBe(0);
    expect(resolveMaxRetries({ max: { lit: 0 } }, undefined)).toBe(0);
    expect(resolveMaxRetries({ max: { lit: -3 } }, undefined)).toBe(0);
    // A malformed descriptor with a non-numeric literal disables, not crashes.
    expect(resolveMaxRetries({ max: { lit: "3" as unknown as number } }, undefined)).toBe(0);
  });

  it("resolves the retry maximum through a ref into the client values", () => {
    expect(resolveMaxRetries({ max: { ref: "m" } }, { m: 3 })).toBe(3);
    expect(resolveMaxRetries({ max: { ref: "m" } }, { m: "3" })).toBe(0);
    expect(resolveMaxRetries({ max: { ref: "m" } }, { m: Infinity })).toBe(0);
    expect(resolveMaxRetries({ max: { ref: "absent" } }, {})).toBe(0);
    expect(resolveMaxRetries({ max: { ref: "m" } }, undefined)).toBe(0);
  });

  it("resolves the timeout, treating absence as none", () => {
    expect(resolveTimeoutMs(undefined, undefined)).toBe(0);
    expect(resolveTimeoutMs(null, undefined)).toBe(0);
    expect(resolveTimeoutMs({ lit: 1500 }, undefined)).toBe(1500);
    expect(resolveTimeoutMs({ ref: "t" }, { t: 250 })).toBe(250);
    expect(resolveTimeoutMs({ ref: "absent" }, {})).toBe(0);
  });
});

describe("backoff", () => {
  it("grows exponentially from the base and caps, scaled by the jitter", () => {
    expect(backoffDelayMs(0, 0.5)).toBe(50);
    expect(backoffDelayMs(1, 0.5)).toBe(100);
    expect(backoffDelayMs(2, 0.5)).toBe(200);
    expect(backoffDelayMs(3, 0.5)).toBe(400);
    expect(backoffDelayMs(4, 0.5)).toBe(800);
    // 100 * 2^5 = 3200 caps at 2000.
    expect(backoffDelayMs(5, 0.5)).toBe(1000);
    expect(backoffDelayMs(9, 1)).toBe(2000);
    expect(backoffDelayMs(3, 0)).toBe(0);
  });
});

describe("retryable classification", () => {
  // Stands in for a generated decode<Op>Error + retryable() pair: the first
  // status match with an agreeing code decides (a declared code must equal
  // the body's "code" field; a null code matches any body).
  function classifyLikeDiscriminator(
    declared: ReadonlyArray<readonly [number, string | null, boolean]>,
  ): (status: number, body: string) => boolean {
    return (status, body) => {
      let code: unknown;
      try {
        const parsed: unknown = JSON.parse(body);
        code = parsed !== null && typeof parsed === "object" ? (parsed as Record<string, unknown>)["code"] : undefined;
      } catch {
        code = undefined;
      }
      for (const [declaredStatus, declaredCode, retryable] of declared) {
        if (declaredStatus !== status) continue;
        if (declaredCode === null || code === declaredCode) return retryable;
      }
      return false;
    };
  }

  const retryable = classifyLikeDiscriminator([
    [429, "overloaded", true],
    [429, "quota", false],
    [503, null, true],
    [404, null, false],
  ]);
  const error = (status: number, body: string): Outcome => ({ outcome: "error", status, body });

  it("always retries a transport failure and never a success", () => {
    expect(isRetryable({ outcome: "transport", cause: new Error("x") }, retryable)).toBe(true);
    expect(isRetryable({ outcome: "success", status: 200, body: "{}" }, retryable)).toBe(false);
    // Even when a success status collides with a declared retryable error
    // (a declared non-2xx success), the outcome kind decides.
    expect(
      isRetryable({ outcome: "success", status: 429, body: '{"code":"overloaded"}' }, retryable),
    ).toBe(false);
  });

  it("discriminates declared errors by status and code", () => {
    expect(isRetryable(error(429, '{"code":"overloaded"}'), retryable)).toBe(true);
    expect(isRetryable(error(429, '{"code":"quota"}'), retryable)).toBe(false);
    expect(isRetryable(error(429, '{"code":"other"}'), retryable)).toBe(false);
    expect(isRetryable(error(429, "not json"), retryable)).toBe(false);
    expect(isRetryable(error(429, "null"), retryable)).toBe(false);
    expect(isRetryable(error(429, '{"code":5}'), retryable)).toBe(false);
    expect(isRetryable(error(503, "not json"), retryable)).toBe(true);
    expect(isRetryable(error(404, "{}"), retryable)).toBe(false);
    expect(isRetryable(error(500, "{}"), retryable)).toBe(false);
  });

  it("treats an absent predicate (no declared errors) as never retryable", () => {
    expect(isRetryable(error(503, "{}"), undefined)).toBe(false);
  });
});

describe("transport slot exclusivity", () => {
  it("rejects both slots set, at assertion and at execute", async () => {
    const { transport } = scripted([{ kind: "response", status: 200, body: "{}" }]);
    const options = { baseUrl: "https://api.test", fetch, transport };
    expect(() => assertExclusiveTransport(options)).toThrow(/mutually exclusive/);
    await expect(execute(descriptor(), {}, options)).rejects.toThrow(/mutually exclusive/);
  });

  it("accepts each slot alone", () => {
    const { transport } = scripted([]);
    expect(() => assertExclusiveTransport({ baseUrl: "b", transport })).not.toThrow();
    expect(() => assertExclusiveTransport({ baseUrl: "b", fetch })).not.toThrow();
    expect(() => assertExclusiveTransport({ baseUrl: "b" })).not.toThrow();
  });
});

describe("canonical transport slot", () => {
  it("receives the built request and its response is classified as usual", async () => {
    let seen: unknown;
    const transport: CanonicalTransport = (req) => {
      seen = req;
      return Promise.resolve({ status: 200, headers: { "x-request-id": "r1" }, body: '{"id":"x"}' });
    };
    const outcome = await execute(
      descriptor({
        http_method: "PUT",
        uri: "/x/{id}",
        bindings: [
          ["id", { kind: "label" }],
          ["field", { kind: "body" }],
        ],
        response_bindings: [["requestId", { kind: "header", name: "X-Request-Id" }]],
      }),
      { id: "1", field: "4" },
      { baseUrl: "https://api.test", transport, headers: { Authorization: "Bearer t" } },
    );
    expect(seen).toEqual({
      method: "PUT",
      url: "https://api.test/x/1",
      headers: { Authorization: "Bearer t", "content-type": "application/json" },
      body: '{"field":"4"}',
    });
    expect(outcome).toEqual({
      outcome: "success",
      status: 200,
      body: JSON.stringify({ id: "x", requestId: "r1" }),
    });
  });

  it("reports a canonical transport rejection as a transport outcome", async () => {
    const boom = new Error("down");
    const transport: CanonicalTransport = () => Promise.reject(boom);
    const outcome = await execute(descriptor(), {}, { baseUrl: "https://api.test", transport });
    expect(outcome).toEqual({ outcome: "transport", cause: boom });
  });

  it("passes no signal when the descriptor declares no timeout", async () => {
    let sawSignal: AbortSignal | undefined = undefined;
    const transport: CanonicalTransport = (_req, signal) => {
      sawSignal = signal;
      return Promise.resolve({ status: 200, headers: {}, body: "{}" });
    };
    await execute(descriptor(), {}, { baseUrl: "https://api.test", transport });
    expect(sawSignal).toBeUndefined();
  });
});

describe("retry loop", () => {
  it("honors the maximum and the backoff sequence", async () => {
    const { transport, attempts } = scripted([
      { kind: "transport_failure" },
      { kind: "transport_failure" },
      { kind: "transport_failure" },
      { kind: "transport_failure" },
    ]);
    const { delays, seams: s } = seams();
    const outcome = await execute(
      descriptor({ retry: { max: { lit: 3 } } }),
      {},
      { baseUrl: "https://api.test", transport },
      undefined,
      undefined,
      s,
    );
    expect(outcome.outcome).toBe("transport");
    expect(attempts()).toBe(4);
    expect(delays).toEqual([50, 100, 200]);
  });

  it("stops at the first non-retryable outcome", async () => {
    const { transport, attempts } = scripted([
      { kind: "response", status: 503, body: "{}" },
      { kind: "response", status: 404, body: "{}" },
    ]);
    const { delays, seams: s } = seams();
    const outcome = await execute(
      descriptor({ retry: { max: { lit: 5 } } }),
      {},
      { baseUrl: "https://api.test", transport },
      undefined,
      (status) => status === 503,
      s,
    );
    expect(outcome).toEqual({ outcome: "error", status: 404, body: "{}" });
    expect(attempts()).toBe(2);
    expect(delays).toEqual([50]);
  });

  it("retries on the native slot too", async () => {
    let calls = 0;
    const fetchImpl = (async () => {
      calls += 1;
      if (calls === 1) throw new Error("down");
      return new Response("{}", { status: 200 });
    }) as unknown as typeof fetch;
    const { seams: s } = seams();
    const outcome = await execute(
      descriptor({ retry: { max: { lit: 1 } } }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
      undefined,
      undefined,
      s,
    );
    expect(outcome.outcome).toBe("success");
    expect(calls).toBe(2);
  });

  it("really sleeps the backoff delay when no seams are given", async () => {
    vi.useFakeTimers();
    vi.spyOn(Math, "random").mockReturnValue(0.5);
    const { transport, attempts } = scripted([
      { kind: "transport_failure" },
      { kind: "response", status: 200, body: "{}" },
    ]);
    let settled = false;
    const pending = execute(
      descriptor({ retry: { max: { lit: 1 } } }),
      {},
      { baseUrl: "https://api.test", transport },
    ).then((outcome) => {
      settled = true;
      return outcome;
    });
    // The first attempt failed and the retry is waiting out its 50ms backoff
    // (0.5 * 100): one tick short of it, nothing has settled.
    await vi.advanceTimersByTimeAsync(49);
    expect(settled).toBe(false);
    await vi.advanceTimersByTimeAsync(1);
    const outcome = await pending;
    expect(outcome.outcome).toBe("success");
    expect(attempts()).toBe(2);
  });

  it("propagates a before_request hook throw without attempting or retrying", async () => {
    const { transport, attempts } = scripted([{ kind: "response", status: 200, body: "{}" }]);
    const boom = new Error("hook broke");
    await expect(
      execute(
        descriptor({ retry: { max: { lit: 3 } } }),
        {},
        { baseUrl: "https://api.test", transport },
        {
          before_request: () => {
            throw boom;
          },
        },
        undefined,
        seams().seams,
      ),
    ).rejects.toBe(boom);
    expect(attempts()).toBe(0);
  });

  it("propagates an after_response hook throw instead of retrying a retryable response", async () => {
    const { transport, attempts } = scripted([{ kind: "response", status: 503, body: "{}" }]);
    const boom = new Error("hook broke");
    await expect(
      execute(
        descriptor({ retry: { max: { lit: 3 } } }),
        {},
        { baseUrl: "https://api.test", transport },
        {
          after_response: () => {
            throw boom;
          },
        },
        (status) => status === 503,
        seams().seams,
      ),
    ).rejects.toBe(boom);
    expect(attempts()).toBe(1);
  });

  it("classifies for retry on the post-hook response", async () => {
    const { transport, attempts } = scripted([{ kind: "response", status: 503, body: "{}" }]);
    const { delays, seams: s } = seams();
    const outcome = await execute(
      descriptor({ retry: { max: { lit: 3 } } }),
      {},
      { baseUrl: "https://api.test", transport },
      { after_response: (res) => ({ ...res, status: 200 }) },
      (status) => status === 503,
      s,
    );
    expect(outcome.outcome).toBe("success");
    expect(attempts()).toBe(1);
    expect(delays).toEqual([]);
  });
});

describe("per-attempt timeout", () => {
  it("times out a hanging canonical transport, aborting the signal", async () => {
    const { transport, attempts } = scripted([
      { kind: "hang" },
      { kind: "response", status: 200, body: "{}" },
    ]);
    const { delays, seams: s } = seams();
    const outcome = await execute(
      descriptor({ retry: { max: { lit: 1 } }, timeout: { lit: 30 } }),
      {},
      { baseUrl: "https://api.test", transport },
      undefined,
      undefined,
      s,
    );
    expect(outcome.outcome).toBe("success");
    expect(attempts()).toBe(2);
    expect(delays).toEqual([50]);
  });

  it("surfaces a timeout without retry as a transport outcome naming the budget", async () => {
    const { transport } = scripted([{ kind: "hang" }]);
    const outcome = await execute(
      descriptor({ timeout: { lit: 30 } }),
      {},
      { baseUrl: "https://api.test", transport },
    );
    expect(outcome.outcome).toBe("transport");
    expect((outcome as { cause: Error }).cause.message).toBe("attempt timed out after 30ms");
  });

  it("times out a transport that ignores the signal entirely", async () => {
    const transport: CanonicalTransport = () => new Promise<CanonicalResponse>(() => {});
    const outcome = await execute(
      descriptor({ timeout: { lit: 20 } }),
      {},
      { baseUrl: "https://api.test", transport },
    );
    expect(outcome.outcome).toBe("transport");
  });

  it("hands the signal to the native fetch when a timeout is declared", async () => {
    let sawSignal: unknown;
    const fetchImpl = (async (_url: unknown, init?: RequestInit) => {
      sawSignal = init?.signal;
      return new Response("{}", { status: 200 });
    }) as unknown as typeof fetch;
    const outcome = await execute(
      descriptor({ timeout: { lit: 1000 } }),
      {},
      { baseUrl: "https://api.test", fetch: fetchImpl },
    );
    expect(outcome.outcome).toBe("success");
    expect(sawSignal).toBeInstanceOf(AbortSignal);
  });

  it("does not time out an attempt that answers within the budget", async () => {
    const { transport } = scripted([{ kind: "response", status: 200, body: "{}" }]);
    const outcome = await execute(
      descriptor({ timeout: { lit: 1000 } }),
      {},
      { baseUrl: "https://api.test", transport },
    );
    expect(outcome).toEqual({ outcome: "success", status: 200, body: "{}" });
  });

  it("clears the expiry timer once the attempt answers", async () => {
    vi.useFakeTimers();
    const { transport } = scripted([{ kind: "response", status: 200, body: "{}" }]);
    const outcome = await execute(
      descriptor({ timeout: { lit: 1000 } }),
      {},
      { baseUrl: "https://api.test", transport },
    );
    expect(outcome.outcome).toBe("success");
    expect(vi.getTimerCount()).toBe(0);
  });
});

File: runtimes/http-go/retry_test.go
package tonohttp

import (
	"context"
	"encoding/json"
	"errors"
	"testing"
	"time"
)

func lit(f float64) *float64 { return &f }
func ref(s string) *string   { return &s }

// A runtime with deterministic seams: jitter pinned to 0.5 and a recording,
// instant sleep.
func deterministic(t *testing.T, opts Options) (*Runtime, *[]time.Duration) {
	t.Helper()
	r := newRuntime(t, opts)
	var delays []time.Duration
	r.randFloat = func() float64 { return 0.5 }
	r.sleep = func(ctx context.Context, d time.Duration) error {
		delays = append(delays, d)
		return nil
	}
	return r, &delays
}

func TestResolveMaxRetries(t *testing.T) {
	values := map[string]any{"f": float64(2), "i": int(3), "i64": int64(4), "s": "5"}
	cases := []struct {
		name string
		spec *RetrySpec
		want int
	}{
		{"nil spec", nil, 0},
		{"literal", &RetrySpec{Max: ValueSource{Lit: lit(2)}}, 2},
		{"fractional floors", &RetrySpec{Max: ValueSource{Lit: lit(2.9)}}, 2},
		{"below one disables", &RetrySpec{Max: ValueSource{Lit: lit(0.5)}}, 0},
		{"zero", &RetrySpec{Max: ValueSource{Lit: lit(0)}}, 0},
		{"negative", &RetrySpec{Max: ValueSource{Lit: lit(-3)}}, 0},
		{"ref float", &RetrySpec{Max: ValueSource{Ref: ref("f")}}, 2},
		{"ref int", &RetrySpec{Max: ValueSource{Ref: ref("i")}}, 3},
		{"ref int64", &RetrySpec{Max: ValueSource{Ref: ref("i64")}}, 4},
		{"ref to non-number", &RetrySpec{Max: ValueSource{Ref: ref("s")}}, 0},
		{"ref to nothing", &RetrySpec{Max: ValueSource{Ref: ref("absent")}}, 0},
		{"empty source", &RetrySpec{}, 0},
	}
	for _, c := range cases {
		if got := resolveMaxRetries(c.spec, values); got != c.want {
			t.Errorf("%s: got %d, want %d", c.name, got, c.want)
		}
	}
}

func TestResolveTimeout(t *testing.T) {
	values := map[string]any{"t": float64(250)}
	if got := resolveTimeout(nil, values); got != 0 {
		t.Fatalf("nil source: %v", got)
	}
	if got := resolveTimeout(&ValueSource{Lit: lit(1500)}, values); got != 1500*time.Millisecond {
		t.Fatalf("literal: %v", got)
	}
	if got := resolveTimeout(&ValueSource{Ref: ref("t")}, values); got != 250*time.Millisecond {
		t.Fatalf("ref: %v", got)
	}
	if got := resolveTimeout(&ValueSource{Ref: ref("absent")}, values); got != 0 {
		t.Fatalf("unresolved ref: %v", got)
	}
}

func TestBackoffDelay(t *testing.T) {
	cases := []struct {
		retry  int
		random float64
		want   time.Duration
	}{
		{0, 0.5, 50 * time.Millisecond},
		{1, 0.5, 100 * time.Millisecond},
		{2, 0.5, 200 * time.Millisecond},
		{3, 0.5, 400 * time.Millisecond},
		{4, 0.5, 800 * time.Millisecond},
		// 100 * 2^5 = 3200 caps at 2000.
		{5, 0.5, 1000 * time.Millisecond},
		{9, 1.0, 2000 * time.Millisecond},
		{3, 0, 0},
	}
	for _, c := range cases {
		if got := backoffDelay(c.retry, c.random); got != c.want {
			t.Errorf("backoffDelay(%d, %v) = %v, want %v", c.retry, c.random, got, c.want)
		}
	}
}

// declaredErrorRule is one entry of a stand-in for a generated
// Decode<Op>Error + Retryable() pair: a status, an optional @errorCode
// discriminator, and whether that error is retryable.
type declaredErrorRule struct {
	Status    int
	Code      *string
	Retryable bool
}

// classifyLikeDiscriminator builds the predicate a generated client passes to
// Execute: the first status match with an agreeing code decides (a declared
// code must equal the body's "code" field; a null code matches any body),
// exactly the ordering discrimination_order gives the real emitter.
func classifyLikeDiscriminator(declared []declaredErrorRule) func(status int, body string) bool {
	return func(status int, body string) bool {
		code := bodyCodeForTest(body)
		for _, e := range declared {
			if e.Status != status {
				continue
			}
			if e.Code == nil || (code != nil && *code == *e.Code) {
				return e.Retryable
			}
		}
		return false
	}
}

func bodyCodeForTest(body string) *string {
	var object map[string]any
	if err := json.Unmarshal([]byte(body), &object); err != nil {
		return nil
	}
	if code, ok := object["code"].(string); ok {
		return &code
	}
	return nil
}

func TestIsRetryableClassification(t *testing.T) {
	retryable := classifyLikeDiscriminator([]declaredErrorRule{
		{Status: 429, Code: ref("overloaded"), Retryable: true},
		{Status: 429, Code: ref("quota"), Retryable: false},
		{Status: 503, Code: nil, Retryable: true},
		{Status: 404, Code: nil, Retryable: false},
	})
	if !isRetryable(Outcome{Kind: OutcomeTransport}, retryable) {
		t.Fatal("transport failures always retry")
	}
	if isRetryable(Outcome{Kind: OutcomeSuccess, Status: 200}, retryable) {
		t.Fatal("a success never retries")
	}
	if !isRetryable(Outcome{Kind: OutcomeError, Status: 429, Body: `{"code":"overloaded"}`}, retryable) {
		t.Fatal("retryable code match must retry")
	}
	if isRetryable(Outcome{Kind: OutcomeError, Status: 429, Body: `{"code":"quota"}`}, retryable) {
		t.Fatal("non-retryable code match must not retry")
	}
	if isRetryable(Outcome{Kind: OutcomeError, Status: 429, Body: `{"code":"other"}`}, retryable) {
		t.Fatal("an unmatched code must not retry")
	}
	if isRetryable(Outcome{Kind: OutcomeError, Status: 429, Body: "not json"}, retryable) {
		t.Fatal("a body without a code cannot match a coded error")
	}
	if !isRetryable(Outcome{Kind: OutcomeError, Status: 503, Body: "not json"}, retryable) {
		t.Fatal("a null code matches any body")
	}
	if isRetryable(Outcome{Kind: OutcomeError, Status: 404, Body: "{}"}, retryable) {
		t.Fatal("declared non-retryable must not retry")
	}
	if isRetryable(Outcome{Kind: OutcomeError, Status: 500, Body: "{}"}, retryable) {
		t.Fatal("an undeclared status must not retry")
	}
	if isRetryable(Outcome{Kind: OutcomeError, Status: 429, Body: `{"code":5}`}, retryable) {
		t.Fatal("a non-string code field cannot match")
	}
	if isRetryable(Outcome{Kind: OutcomeError, Status: 500, Body: "{}"}, nil) {
		t.Fatal("a nil predicate (no declared errors) must not retry")
	}
}

func TestRetryLoopHonorsMaxAndBackoff(t *testing.T) {
	attempts := 0
	transport := func(ctx context.Context, req CanonicalRequest) (CanonicalResponse, error) {
		attempts++
		return CanonicalResponse{}, errors.New("down")
	}
	r, delays := deterministic(t, Options{BaseURL: "https://api.test", Transport: transport})
	d := desc(func(d *WireDescriptor) { d.Retry = &RetrySpec{Max: ValueSource{Lit: lit(3)}} })
	outcome, err := r.Execute(context.Background(), d, nil, nil, nil)
	if err != nil || outcome.Kind != OutcomeTransport {
		t.Fatalf("exhausted retries must return the last outcome: %+v %v", outcome, err)
	}
	if attempts != 4 {
		t.Fatalf("attempts: %d", attempts)
	}
	want := []time.Duration{50 * time.Millisecond, 100 * time.Millisecond, 200 * time.Millisecond}
	if len(*delays) != len(want) {
		t.Fatalf("delays: %v", *delays)
	}
	for i, w := range want {
		if (*delays)[i] != w {
			t.Fatalf("delay %d: %v, want %v", i, (*delays)[i], w)
		}
	}
}

func TestRetryStopsAtFirstNonRetryableOutcome(t *testing.T) {
	responses := []CanonicalResponse{
		{Status: 503, Body: `{}`},
		{Status: 404, Body: `{}`},
		{Status: 200, Body: `{}`},
	}
	attempts := 0
	transport := func(ctx context.Context, req CanonicalRequest) (CanonicalResponse, error) {
		res := responses[attempts]
		attempts++
		return res, nil
	}
	r, delays := deterministic(t, Options{BaseURL: "https://api.test", Transport: transport})
	d := desc(func(d *WireDescriptor) {
		d.Retry = &RetrySpec{Max: ValueSource{Lit: lit(5)}}
	})
	retryable := func(status int, body string) bool { return status == 503 }
	outcome, err := r.Execute(context.Background(), d, nil, nil, retryable)
	if err != nil || outcome.Kind != OutcomeError || outcome.Status != 404 {
		t.Fatalf("first non-retryable outcome must surface: %+v %v", outcome, err)
	}
	if attempts != 2 || len(*delays) != 1 {
		t.Fatalf("attempts %d, delays %v", attempts, *delays)
	}
}

func TestPerAttemptTimeoutRetriesAndRecovers(t *testing.T) {
	attempts := 0
	transport := func(ctx context.Context, req CanonicalRequest) (CanonicalResponse, error) {
		attempts++
		if attempts == 1 {
			// Honor the ctx deadline, as the canonical transport contract
			// asks; bail out fast if a mutant keeps the deadline from firing.
			select {
			case <-ctx.Done():
				return CanonicalResponse{}, ctx.Err()
			case <-time.After(2 * time.Second):
				return CanonicalResponse{}, errors.New("the per-attempt deadline never fired")
			}
		}
		return CanonicalResponse{Status: 200, Body: "{}"}, nil
	}
	r, delays := deterministic(t, Options{BaseURL: "https://api.test", Transport: transport})
	d := desc(func(d *WireDescriptor) {
		d.Retry = &RetrySpec{Max: ValueSource{Lit: lit(1)}}
		d.Timeout = &ValueSource{Lit: lit(20)}
	})
	start := time.Now()
	outcome, err := r.Execute(context.Background(), d, nil, nil, nil)
	if err != nil || outcome.Kind != OutcomeSuccess {
		t.Fatalf("timeout must count as transport failure and retry: %+v %v", outcome, err)
	}
	if attempts != 2 || len(*delays) != 1 {
		t.Fatalf("attempts %d, delays %v", attempts, *delays)
	}
	if elapsed := time.Since(start); elapsed < 20*time.Millisecond {
		t.Fatalf("first attempt returned before its deadline: %v", elapsed)
	}
}

func TestCancellationDuringBackoffSurfacesAsTransport(t *testing.T) {
	transport := func(ctx context.Context, req CanonicalRequest) (CanonicalResponse, error) {
		return CanonicalResponse{}, errors.New("down")
	}
	r := newRuntime(t, Options{BaseURL: "https://api.test", Transport: transport})
	r.randFloat = func() float64 { return 0.5 }
	ctx, cancel := context.WithCancel(context.Background())
	r.sleep = func(ctx context.Context, d time.Duration) error {
		cancel()
		return sleepContext(ctx, d)
	}
	d := desc(func(d *WireDescriptor) { d.Retry = &RetrySpec{Max: ValueSource{Lit: lit(3)}} })
	outcome, err := r.Execute(ctx, d, nil, nil, nil)
	if err != nil || outcome.Kind != OutcomeTransport || !errors.Is(outcome.Cause, context.Canceled) {
		t.Fatalf("cancellation during backoff: %+v %v", outcome, err)
	}
}

func TestSleepContext(t *testing.T) {
	if err := sleepContext(context.Background(), time.Millisecond); err != nil {
		t.Fatalf("normal sleep: %v", err)
	}
	ctx, cancel := context.WithCancel(context.Background())
	cancel()
	// Half a second, not an hour: a mutant that ignores the cancellation then
	// fails fast on the return value instead of hanging the binary.
	if err := sleepContext(ctx, 500*time.Millisecond); !errors.Is(err, context.Canceled) {
		t.Fatalf("cancelled sleep: %v", err)
	}
}
